### PR TITLE
Bytecode Instrumentation - Pass parameters by ref

### LINKED ISSUE #2424: Cleanup unused bytecode methods (without ref)
Follow up to: Update native code #2280

>What is the reason to keep old (without ref) version of the methods?

_Originally posted by @Kielek in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2280#discussion_r1145919182_

### LINKED ISSUE #2427: Native code - enable by ref by default
Probably separate GH issue:
Can it be enabled by default? I do not see any condition when it should be disabled.

_Originally posted by @Kielek in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2280#discussion_r1156999071_

### PR DESCRIPTION
**Pull request by zacharycmontoya**
## Why

In the recent native code update (https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2280), the following optimizations were added:
- Passing the method parameters for the bytecode-instrumented method ByRef into the `CallTargetInvoker.BeginMethod` API
- Passing the `CallTargetState` parameter ByRef into the `CallTargetInvoker.EndMethod` API

This PR removes the non-ByRef APIs and unused codepaths, since there is only performance upside to passing ByRef and there are no backward compatibility concerns that require us to keep the previous non-byref method signatures.

Note: In some parts of this PR, the changes may keep some existing `enable_by_ref_instrumentation`/`enable_calltarget_state_by_ref` variables and conditional branches intact, to make it easier to adopt upstream updates (similar code structure). However, if we'd like to remove those branches entirely since they will never be called, we can do so.

Fixes #2424, #2427

## What

- Removes `EnableByRefInstrumentation` from the native API and defaults `enable_by_ref_instrumentation` to true throughout the native code
- Removes `EnableCallTargetStateByRef` from the native API and defaults `enable_calltarget_state_by_ref` to true throughout the native code

## Tests

Runs the existing Workflow

## Checklist

- ~[ ] `CHANGELOG.md` is updated.~
- ~[ ] Documentation is updated.~
- ~[ ] New features are covered by tests.~

### CODE REVIEW

File: src/OpenTelemetry.AutoInstrumentation.Native/calltarget_tokens.h [hunk 1]
```diff
@@ -31,8 +31,8 @@ class CallTargetTokens
 {
 private:
     ModuleMetadata* module_metadata_ptr = nullptr;
-    const bool enable_by_ref_instrumentation = false;
-    const bool enable_calltarget_state_by_ref = false;
+    const bool enable_by_ref_instrumentation = true;
```
Comment (Kielek) [inline src/OpenTelemetry.AutoInstrumentation.Native/calltarget_tokens.h:34]: Could you please add comment that we decided to keep it to make upstream sync easier?

File: src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.h [hunk 2]
```diff
@@ -52,8 +52,8 @@ class CorProfiler : public CorProfilerBase
     // CallTarget Members
     //
     std::shared_ptr<RejitHandler> rejit_handler = nullptr;
-    bool enable_by_ref_instrumentation = false;
-    bool enable_calltarget_state_by_ref = false;
+    bool enable_by_ref_instrumentation = true;
```
Comment (Kielek) [inline src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.h:55]: Same here.

**Review by Kielek — APPROVED** [review]
I would like to see some additional comments in the code + fix native code formating.

Other parts looks good.

File: src/OpenTelemetry.AutoInstrumentation.Native/calltarget_tokens.h (hunk 1 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation.Native/calltarget_tokens.h:37]: Added in bc10d076daf14543982643b5269c9566b811d6ee

File: src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.h (hunk 2 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.h:58]: Added in bc10d076daf14543982643b5269c9566b811d6ee

File: src/OpenTelemetry.AutoInstrumentation.Native/calltarget_tokens.h (hunk 1 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation.Native/calltarget_tokens.h:37]: Added in bc10d076daf14543982643b5269c9566b811d6ee

**Review by pjanotti — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

**Pull request merged**